File: web/lib/types/parse.ts
// NLP parser API contract.
// MIRRORS ai-services/app/models/parse.py (the canonical source). Any change to one
// must update the other in the SAME commit (see CLAUDE.md > Rules).

export type NeedsSeverity = "low" | "moderate" | "high" | "critical";
export type RoadStatus = "passable" | "impassable" | "unknown";
export type ReportStatus = "pending" | "flagged";
export type ParseProvider = "sea-lion" | "gemini";

export interface ParseRequest {
  /** Raw field-report text (Bisaya/Tagalog/EN). */
  text: string;
  /** Caller key, echoed back. */
  id?: string | null;
}

/** Per-field values with their individual confidences (for the coordinator UI). */
export interface Extraction {
  location: string | null;
  location_confidence: number;
  population_estimate: number | null;
  population_confidence: number;
  needs_severity: NeedsSeverity | null;
  needs_severity_confidence: number;
  road_status: RoadStatus;
  road_status_confidence: number;
}

/** A field_reports insert payload (text-derived fields only). */
export interface NormalizedFieldReport {
  source: "parsed";
  raw_text: string;
  /** Extracted place name; geocoded to a barangay downstream. */
  location_text: string | null;
  population_estimate: number | null;
  needs_severity: NeedsSeverity | null;
  road_status: RoadStatus;
  road_impassable: boolean;
  /** Overall extraction confidence, [0,1]. */
  confidence: number;
  /** "flagged" (below threshold, needs review) or "pending". */
  status: ReportStatus;
}

export interface ParseResponse {
  field_report: NormalizedFieldReport;
  extraction: Extraction;
  provider: ParseProvider;
  /** True => below threshold, not auto-committed. */
  needs_review: boolean;
  latency_ms: number;
  /** English translation of raw_text; null if already English. */
  translated_text: string | null;
  id?: string | null;
}

export interface TranslateRequest {
  /** Report text to translate to English. */
  text: string;
  /** Caller key, echoed back. */
  id?: string | null;
}

export interface TranslateResponse {
  /** English translation; null if the text is already English. */
  translated_text: string | null;
  provider: ParseProvider | null;
  latency_ms: number;
  id?: string | null;
}

File: ai-services/app/models/parse.py
"""API contracts for the NLP parser — SOURCE OF TRUTH (mirror in web/lib/types).

Turns unstructured Filipino/Bisaya/Tagalog field-report text into a normalized
field_reports payload. `NormalizedFieldReport` mirrors the public.field_reports insert
columns (Phase 0.2): the parser fills the text-derivable fields; barangay matching and
geocoding the `location_text` happen downstream (pipeline 4.1).
"""
from typing import Literal, Optional

from pydantic import BaseModel, Field

NeedsSeverity = Literal["low", "moderate", "high", "critical"]
RoadStatus = Literal["passable", "impassable", "unknown"]
ReportStatus = Literal["pending", "flagged"]


class ParseRequest(BaseModel):
    text: str = Field(..., min_length=1, description="raw field-report text (Bisaya/Tagalog/EN)")
    id: Optional[str] = Field(None, description="caller key, echoed back")


class Extraction(BaseModel):
    """Per-field values with their individual confidences (for the coordinator UI)."""

    location: Optional[str] = None
    location_confidence: float = Field(0.0, ge=0.0, le=1.0)
    population_estimate: Optional[int] = Field(None, ge=0)
    population_confidence: float = Field(0.0, ge=0.0, le=1.0)
    needs_severity: Optional[NeedsSeverity] = None
    needs_severity_confidence: float = Field(0.0, ge=0.0, le=1.0)
    road_status: RoadStatus = "unknown"
    road_status_confidence: float = Field(0.0, ge=0.0, le=1.0)


class NormalizedFieldReport(BaseModel):
    """A field_reports insert payload (text-derived fields only)."""

    source: Literal["parsed"] = "parsed"
    raw_text: str
    location_text: Optional[str] = Field(None, description="extracted place name; geocoded downstream")
    population_estimate: Optional[int] = Field(None, ge=0)
    needs_severity: Optional[NeedsSeverity] = None
    road_status: RoadStatus = "unknown"
    road_impassable: bool = False
    confidence: float = Field(..., ge=0.0, le=1.0, description="overall extraction confidence")
    status: ReportStatus = Field(
        ..., description="'flagged' (below threshold, needs review) or 'pending'"
    )


class ParseResponse(BaseModel):
    field_report: NormalizedFieldReport
    extraction: Extraction
    provider: Literal["sea-lion", "gemini"]
    needs_review: bool = Field(..., description="True => below threshold, not auto-committed")
    latency_ms: float
    translated_text: Optional[str] = Field(
        None, description="English translation of raw_text; null if already English"
    )
    id: Optional[str] = None


class TranslateRequest(BaseModel):
    text: str = Field(..., min_length=1, description="report text to translate to English")
    id: Optional[str] = Field(None, description="caller key, echoed back")


class TranslateResponse(BaseModel):
    translated_text: Optional[str] = Field(
        None, description="English translation; null if the text is already English"
    )
    provider: Optional[Literal["sea-lion", "gemini"]] = None
    latency_ms: float
    id: Optional[str] = None
